lib/AST/ASTContext.cpp

```diff
@@ -1828,6 +1837,36 @@ GenericSignatureBuilder *ASTContext::getOrCreateGenericSignatureBuilder(
   return builder;
 }
 
+RequirementMachine *ASTContext::getOrCreateRequirementMachine(
+    CanGenericSignature sig) {
+  // Check whether we already have a requirement machine for this
+  // signature.
+  auto arena = getArena(sig);
+  auto &machines = getImpl().getArena(arena).RequirementMachines;
+  auto known = machines.find(sig);
+  if (known != machines.end()) {
+    auto *machine = known->second.get();
+    if (!machine->isComplete()) {
+      llvm::errs() << "Re-entrant construction of requirement "
+                   << "machine for " << sig << "\n";
+      abort();
+    }
+
+    return machine;
+  }
+
+  // Create a new requirement machine with the given signature.
+  auto machine = new RequirementMachine(*this);
+
+  // Store this requirement machine before adding the signature,
+  // to catch re-entrant construction.
+  machines[sig] = std::unique_ptr<RequirementMachine>(machine);
```

Comment:
Nit: Could save a lookup by doing a lookup above then emplacing into it if there’s no existing match for the signature.

Reply:
Fixed.